### DISCUSSION TITLE
Harmonize add/replace queue options

### DISCUSSION THREAD

**Discussion by tsunulukai** [jcorporation/myMPD#534] [Ideas]
Throughout the interface, depending on the location, we get different options to add/replace songs/albums/playlists in the queue/playlist/homescreen.
Although it makes sense for some of those options to offer extra choices, the standard set of options could be harmonized.

Those locations are:
- Settings / Advanced / Default click options
- Browser --> Song Options (button at the right)
- Playlist --> Playlist options
- Home icon / Edit / Action

It would be great if all those different locations could offer the same options where it makes sense to do so. Here's a tabular overview of how this could be streamlined.

| Option | Default click options | Queue | Browser (DB / FS) & Search| Playlist | Home icon action
| --- | :---: | :---: | :---:| :---:| :---:|
| Append to queue | V | X | V | V | V |
| Append to queue and play  | V | X | V | V | V |
| Insert in queue after current song  | V | X | V | V | V |
| Replace queue  | V | X | V | V | V |
| Replace queue and play  | V | X | V | V | V |
| Add to playlist  | X | V | V | V | X |
| Add to homescreen  | X | V | V | V | X |
| Queue options (Remove (Up-/Down- wards), Play after current song) | X | V | X | X | X |
| Song Details | X | V | V | V | X |

Notes: 
- Add to homescreen should be available for playlist/individual songs/albums/folders - curently only available for playlists, although the homescreen supports the types "plist" and "song" ). t's already possible by manually editing the home_list state file with a "song" entry.
- Queue actions should be the same for individual songs as for albums or playlists; I don't see any reason to treat grouped or individual items differently when appending/replacing the queue.
- Append and Replace queue items could exist as they are or with a variant "and play", which would basically directly jump to the newly added entry and play it directly.

What's your take on this ?

**Comment by jcorporation** [comment]
I agree, harmonized options would be perfect. 

Unfortunately, not all options are available through the mpd protocol. 

Add after current song is not available for playlists or search results. For add and play it is the same. For some missing protocol features there already workarounds in place, e. g. for replace queue.

**Reply by tsunulukai** [reply]
MPDroid (https://github.com/abarisain/dmix) implements "add and play" for individual songs as well as playlists or folders.

It does so by sending two distinct commands to the MPD daemon:
```
add <folder>
play "<queueid>"
```
So I guess the tracking of the queue size is done on the client to allow to to know where to jump after it added the new song(s) to the queue.

**Reply by jcorporation** [reply]
There is a addid command that works for folders, songs and urls but not for mpd playlists. For other playlists this command should work.

Tracking the queue size is no good idea: what is if the queue gets changed between add and play or get queue length and add?

**Reply by tsunulukai** [reply]
It should not happen when a single user is commandeering mpd... but if multiple users are sending commands simultaneously, there could indeed be race conditions where this would fail to behave properly.
Never had the issue durings years of mpdroid use though. 

in the online protocol documentation (https://mpd.readthedocs.io/en/latest/protocol.html), addid doesn't seem to accept anything else than a single file ("non-recursive") :

```
add {URI}
    Adds the file URI to the playlist (directories add recursively). URI can also be a single file.

addid {URI} [POSITION]
    Adds a song to the playlist (non-recursive) and returns the song id. URI is always a single file or URL
```

This seem confirmed by manual interaction with mpd:
```
# telnet 127.0.0.1 6600
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
OK MPD 0.22.0
password <password>
OK
add Pop/_Misc
OK
addid Pop/_Misc
ACK [50@0] {addid} No such song
addid "Pop/_Misc/testsong.mp3"
Id: 167
OK
```

So I guess handling this can either be done with the add / play combo.
the song ID to play could be deduced from the playlist command, either
- by running it before the add and hoping the race condition you mentioned above doesn't happen; the song id is then the last returned by the command +1.
- by running between add & play, and searching for the first file of the "selection" we just added, starting from the end of the playlist. but once again, we could hit a race condition under certain circumstances.

Not sure how to best handle this. In an ideal world, addid would work recursively and return an ID-range if when multiple songs get added, but as you mentioned, we have to deal with what the protocol specs offer us today...

**Reply by jcorporation** [reply]
I must look if myMPD tries addid for folders…

I think it could occur if the consume mode manipulates the queue or an another mpd client. All requests inside myMPD are serialised. But I think this needs further investigation.

However:
I do not want implement functions that have race conditions. If there is no proper workaround, a issue for protocol extension should be created.

**Comment by jcorporation** [comment]
I looked through the actual implementation and the mpd protocol details. I hope I don't missed anything.

Already implemented:
- Append to queue
- Replace queue
- Replace queue and play (works with autoplay enabled)
- Queue options (Remove (Up-/Down- wards), Play after current song
- Song Details 

Missing in queue view:
- Add to playlist -> implemented in the v8.1.0 branch

Only for single songs possible
- Append to queue and play
- Insert in queue after current song

Must be manually implemented -> get individual songs -> add it one by one (command list) -> play first song id

ToDo:
- Add to homescreen  for 
  - Items: songs
    - Actions: Append to queue, Append to queue and play, Insert in queue after current song
  - Items: albums, folders, streams
    - Actions: Append to queue
  - Items: playlist
    -  Actions: append to queue  -> implemented in the v8.1.0 branch
- Add "... and play" options

**Comment by jcorporation** [comment]
https://github.com/MusicPlayerDaemon/MPD/issues/888

This new feature helps implementing "Insert in queue after current song" for search results, albums and playlists.
